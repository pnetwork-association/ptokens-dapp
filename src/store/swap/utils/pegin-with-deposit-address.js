import {
  showDepositAddressModal,
  hideDepositAddressModal,
  updateProgress,
  loadBalanceByAssetId,
  resetProgress,
  updateSwapButton
} from '../swap.actions'
import { toastr } from 'react-redux-toastr'
import { getCorrespondingBaseTxExplorerLink } from '../../../utils/ptokens-sm-utils'

let promiEvent = null

const peginWithDepositAddress = async ({ ptokens, address, ptoken, dispatch }) => {
  if (promiEvent) {
    promiEvent.removeAllListeners()
  }

  let depositAddress = null
  try {
    depositAddress = await ptokens[ptoken.workingName].getDepositAddress(address)
  } catch (_err) {
    dispatch(resetProgress())
    return
  }

  dispatch(updateSwapButton('Swapping ...', true))

  dispatch(
    updateProgress({
      show: true,
      percent: 0,
      message: 'Waiting for a deposit ...',
      steps: [0, 20, 40, 60, 80, 100],
      terminated: false
    })
  )

  dispatch(showDepositAddressModal(ptoken, depositAddress.toString()))

  promiEvent = depositAddress.waitForDeposit()
  promiEvent
    .once('nativeTxBroadcasted', _tx => {
      const nativeTransactionField = {
        btc: 'txid',
        ltc: 'txid',
        doge: 'tx_hash'
      }

      dispatch(hideDepositAddressModal())

      toastr.success('Transaction broadcasted!', 'Click here to see it', {
        timeOut: 0,
        onToastrClick: () =>
          window.open(
            `${getCorrespondingBaseTxExplorerLink(ptoken.id, 'native')}${
              _tx[nativeTransactionField[ptoken.nativeBlockchain.toLowerCase()]]
            }`,
            '_blank'
          )
      })

      dispatch(
        updateProgress({
          show: true,
          percent: 20,
          message: 'Deposit detected! Waiting for confirmation ...',
          steps: [0, 20, 40, 60, 80, 100],
          terminated: false
        })
      )
    })
    .once('nativeTxConfirmed', () => {
      dispatch(
        updateProgress({
          show: true,
          percent: 40,
          message: 'Waiting for the pNetwork to detect your transaction ...',
          steps: [0, 20, 40, 60, 80, 100],
          terminated: false
        })
      )
    })
    .once('nodeReceivedTx', () => {
      dispatch(
        updateProgress({
          show: true,
          percent: 60,
          message: 'Enclave received the transaction, broadcasting ...',
          steps: [0, 20, 40, 60, 80, 100],
          terminated: false
        })
      )
    })
    .once('nodeBroadcastedTx', _report => {
      toastr.success('Transaction broadcasted!', 'Click here to see it', {
        timeOut: 0,
        onToastrClick: () =>
          window.open(`${getCorrespondingBaseTxExplorerLink(ptoken.id, 'host')}${_report.broadcast_tx_hash}`, '_blank')
      })

      dispatch(
        updateProgress({
          show: true,
          percent: 80,
          message: 'Asset swap transaction completed, waiting for confirmation ...',
          steps: [0, 20, 40, 60, 80, 100],
          terminated: false
        })
      )
    })
    .then(_result => {
      dispatch(
        updateProgress({
          show: true,
          percent: 100,
          message: 'Transaction Confirmed.',
          steps: [0, 20, 40, 60, 80, 100],
          terminated: true
        })
      )

      dispatch(updateSwapButton('Get Deposit Address'))
      setTimeout(() => dispatch(loadBalanceByAssetId(ptoken.id)), 2000)
    })
    .catch(_err => {
      dispatch(updateSwapButton('Get Deposit Address'))
      dispatch(hideDepositAddressModal())
      dispatch(resetProgress())
    })
}

export default peginWithDepositAddress
