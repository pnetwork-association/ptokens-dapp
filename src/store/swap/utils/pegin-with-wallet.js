import { getCorrespondingBaseTxExplorerLink } from '../../../utils/ptokens-sm-utils'
import { eth } from 'ptokens-utils'
import Web3 from 'web3'
import ERC20Abi from '../../../utils/abi/ERC20.json'
import BigNumber from 'bignumber.js'
import { updateProgress, loadBalanceByAssetId, resetProgress, updateSwapButton } from '../swap.actions'
import { updateInfoModal } from '../../pages/pages.actions'
import { parseError } from '../../../utils/errors'
import {
  getWalletAccountByBlockchain,
  getWalletPermissionByBlockchain,
  getWalletByBlockchain
} from '../../wallets/wallets.selectors'
import { PUOS_ON_ULTRA_MAINNET } from '../../../constants'

let promiEvent = null

const peginWithWallet = async ({ ptokens, ptoken, params, dispatch }) => {
  if (promiEvent) {
    promiEvent.removeAllListeners()
  }

  let link

  // NOTE: avoids brave metamask gas estimation fails
  params[params.length] = {
    gas: ptoken.gasLimitPegin ? ptoken.gasLimitPegin : 200000,
    blocksBehind: 3,
    expireSeconds: 60,
    permission: getWalletPermissionByBlockchain(ptoken.nativeBlockchain) || 'active',
    actor: getWalletAccountByBlockchain(ptoken.nativeBlockchain)
  }
  params[0] = BigNumber(params[0])
    .multipliedBy(10 ** ptoken.nativeDecimals)
    .toFixed()

  // NOTE: peth uses ethers
  if ((ptoken.isPerc20 && ptoken.name !== 'pETH') || ptoken.isBep20) {
    try {
      await ptokens[ptoken.workingName].select()
      const info = await ptokens[ptoken.workingName].selectedNode.getInfo()
      const { account, provider } = getWalletByBlockchain(ptoken.nativeBlockchain)
      const web3 = new Web3(provider)

      const toApprove = new web3.eth.Contract(ERC20Abi, eth.addHexPrefix(info.native_smart_contract_address))
      const allowance = await toApprove.methods.allowance(account, eth.addHexPrefix(info.native_vault_address)).call()
      if (!BigNumber(allowance).isGreaterThanOrEqualTo(params[0])) {
        const _approve = () =>
          new Promise(_resolve =>
            toApprove.methods
              .approve(eth.addHexPrefix(info.native_vault_address), params[0])
              .send({ from: account, gas: 75000 })
              .once('hash', _hash => {
                link = `${getCorrespondingBaseTxExplorerLink(ptoken.id, 'native')}${_hash}`
              })
              .then(_resolve)
          )

        await _approve()
      }
    } catch (_err) {
      console.error(_err)
      return
    }
  }

  dispatch(
    updateProgress({
      show: true,
      percent: 0,
      message: 'Waiting for the transaction to be signed ...',
      steps: [0, 20, 40, 60, 80, 100],
      terminated: false
    })
  )

  // NOTE: puos on ultra works different than the other ptokens
  if (ptoken.id === PUOS_ON_ULTRA_MAINNET) {
    const web3 = new Web3()
    const metadata = web3.utils.utf8ToHex(params[1])
    params[1] = 'swapcontract'
    params.splice(2, 0, metadata)
    console.log(params)
    return
    //promiEvent = ptokens[ptoken.workingName.toLowerCase()].issueWithMetadata(...params)
  } else {
    // TODO: fix peth since now would be pweth
    promiEvent = ptokens[ptoken.workingName.toLowerCase()].issue(...params)
  }

  promiEvent
    .once('nativeTxBroadcasted', _hash => {
      link = `${getCorrespondingBaseTxExplorerLink(ptoken.id, 'native')}${_hash}`
      dispatch(
        updateProgress({
          show: true,
          percent: 20,
          message: `<a href="${link}" target="_blank">Transaction</a> broadcasted! Waiting for confirmation ...`,
          steps: [0, 20, 40, 60, 80, 100],
          terminated: false
        })
      )
    })
    .once('nativeTxConfirmed', _e => {
      if (
        ptoken.nativeBlockchain === 'EOS' ||
        ptoken.nativeBlockchain === 'TELOS' ||
        ptoken.nativeBlockchain === 'ULTRA'
      ) {
        link = `${getCorrespondingBaseTxExplorerLink(ptoken.id, 'native')}${_e.transaction_id}`
        dispatch(
          updateProgress({
            show: true,
            percent: 20,
            message: `<a href="${link}" target="_blank">Transaction</a> broadcasted! Waiting for confirmation ...`,
            steps: [0, 20, 40, 60, 80, 100],
            terminated: false
          })
        )
      }

      dispatch(
        updateProgress({
          show: true,
          percent: 40,
          message: `Waiting for the pNetwork to detect your <a href="${link}" target="_blank">transaction</a> ...`,
          steps: [0, 20, 40, 60, 80, 100],
          terminated: false
        })
      )
    })
    .once('nodeReceivedTx', () => {
      dispatch(
        updateProgress({
          show: true,
          percent: 60,
          message: `Enclave received the <a href="${link}" target="_blank">transaction</a>, broadcasting ...`,
          steps: [0, 20, 40, 60, 80, 100],
          terminated: false
        })
      )
    })
    .once('nodeBroadcastedTx', _report => {
      link = `${getCorrespondingBaseTxExplorerLink(ptoken.id, 'host')}${_report.broadcast_tx_hash}`
      dispatch(
        updateProgress({
          show: true,
          percent: 80,
          message: `Asset swap <a href="${link}" target="_blank">transaction</a> completed, waiting for confirmation ...`,
          steps: [0, 20, 40, 60, 80, 100],
          terminated: false
        })
      )
    })
    .then(_result => {
      dispatch(
        updateProgress({
          show: true,
          percent: 100,
          message: `<a href="${link}" target="_blank">Transaction</a> Confirmed.`,
          steps: [0, 20, 40, 60, 80, 100],
          terminated: true
        })
      )

      dispatch(updateSwapButton('Swap'))
      // TODO: load balance also for native asset
      setTimeout(() => dispatch(loadBalanceByAssetId(ptoken.id)), 2000)
    })
    .catch(_err => {
      console.error(_err)
      const { showModal } = parseError(_err)
      if (showModal) {
        dispatch(
          updateInfoModal({
            show: true,
            text: 'Error during pegin, try again!',
            icon: 'cancel'
          })
        )
      }
      dispatch(updateSwapButton('Swap'))
      dispatch(resetProgress())
    })
}

export default peginWithWallet
