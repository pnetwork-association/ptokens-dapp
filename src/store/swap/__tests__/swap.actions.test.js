import algosdk from 'algosdk'
import { pTokensAlgorandAsset } from 'ptokens-assets-algorand'
import { pTokensEosioAsset } from 'ptokens-assets-eosio'
import { pTokensSwapBuilder, pTokensSwap } from 'ptokens-swap'
import { vi, describe, expect } from 'vitest'

import assets from '../../../settings/swap-assets'
import * as wallets from '../../wallets/wallets.selectors'
import { swap } from '../swap.actions'
import * as peginWithDepositAddress from '../utils/pegin-with-deposit-address'
import * as peginWithWallet from '../utils/pegin-with-wallet'
import * as pegout from '../utils/pegout'

describe('swap', () => {
  beforeAll(() => {
    vi.mock('ptokens-node')
  })

  beforeEach(() => {
    vi.restoreAllMocks()
  })

  it('Should peg-in with deposit address pBTC on Ethereum', async () => {
    const peginWithDepositAddressSpy = vi.spyOn(peginWithDepositAddress, 'default').mockResolvedValue()
    const addDestinationAssetSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'addDestinationAsset')
    const setAmountSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'setAmount')
    const from = assets.find((_el) => _el.id === 'BTC')
    const to = assets.find((_el) => _el.id === 'PBTC_ON_ETH_MAINNET')
    const amount = '10'
    const address = '0xF39d30Fa570db7940e5b3A3e42694665A1449E4B'
    const f = swap(from, to, amount, address)
    const dispatch = vi.fn()
    await f(dispatch)
    expect(setAmountSpy).toHaveBeenCalledTimes(1)
    expect(setAmountSpy).toHaveBeenNthCalledWith(1, '10')
    expect(addDestinationAssetSpy).toHaveBeenCalledTimes(1)
    expect(addDestinationAssetSpy).toHaveBeenNthCalledWith(
      1,
      expect.anything(),
      '0xF39d30Fa570db7940e5b3A3e42694665A1449E4B'
    )
    expect(peginWithDepositAddressSpy).toHaveBeenCalledTimes(1)
    expect(peginWithDepositAddressSpy).toHaveBeenNthCalledWith(1, {
      swap: expect.any(pTokensSwap),
      ptokenFrom: from,
      ptokenTo: to,
      dispatch,
    })
  })

  it('Should peg-out TLOS on BSC', async () => {
    const pegoutSpy = vi.spyOn(pegout, 'default').mockResolvedValue()
    const addDestinationAssetSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'addDestinationAsset')
    const setAmountSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'setAmount')
    const from = assets.find((_el) => _el.id === 'TLOS_ON_BSC_MAINNET')
    const to = assets.find((_el) => _el.id === 'TLOS')
    const amount = '100'
    const address = '0xF39d30Fa570db7940e5b3A3e42694665A1449E4B'
    const f = swap(from, to, amount, address, { pegoutToTelosEvmAddress: true })
    const dispatch = vi.fn()
    await f(dispatch)
    expect(setAmountSpy).toHaveBeenCalledTimes(1)
    expect(setAmountSpy).toHaveBeenNthCalledWith(1, '100')
    expect(addDestinationAssetSpy).toHaveBeenCalledTimes(1)
    expect(addDestinationAssetSpy).toHaveBeenNthCalledWith(
      1,
      expect.anything(),
      'devm.ptokens',
      '0x307846333964333046613537306462373934306535623341336534323639343636354131343439453442'
    )
    expect(pegoutSpy).toHaveBeenCalledTimes(1)
    expect(pegoutSpy).toHaveBeenNthCalledWith(1, {
      swap: expect.any(pTokensSwap),
      ptokenFrom: from,
      ptokenTo: to,
      dispatch,
    })
  })

  it('Should peg-out pUOS on Ultra', async () => {
    vi.spyOn(wallets, 'getWallets').mockReturnValue({ ultra: { provider: undefined, account: 'sender' }, eth: {} })
    const pegoutSpy = vi.spyOn(pegout, 'default').mockResolvedValue()
    const addDestinationAssetSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'addDestinationAsset')
    const setAmountSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'setAmount')
    const setCustomActionsSpy = vi.spyOn(pTokensEosioAsset.prototype, 'setCustomActions')
    const from = assets.find((_el) => _el.id === 'PUOS_ON_ULTRA_MAINNET')
    const to = assets.find((_el) => _el.id === 'UOS')
    const amount = '100'
    const address = '0xF39d30Fa570db7940e5b3A3e42694665A1449E4B'
    const f = swap(from, to, amount, address)
    const dispatch = vi.fn()
    await f(dispatch)
    expect(setAmountSpy).toHaveBeenCalledTimes(1)
    expect(setAmountSpy).toHaveBeenNthCalledWith(1, '100')
    expect(setCustomActionsSpy).toHaveBeenCalledTimes(1)
    expect(setCustomActionsSpy).toHaveBeenNthCalledWith(1, [
      {
        abi: expect.anything(),
        arguments: {
          from: 'sender',
          memo: '0xF39d30Fa570db7940e5b3A3e42694665A1449E4B',
          quantity: '100.00000000 UOS',
          to: 'ultra.swap',
        },
        contractAddress: 'eosio.token',
        method: 'transfer',
      },
    ])
    expect(addDestinationAssetSpy).toHaveBeenCalledTimes(1)
    expect(addDestinationAssetSpy).toHaveBeenNthCalledWith(
      1,
      expect.anything(),
      '0xF39d30Fa570db7940e5b3A3e42694665A1449E4B'
    )
    expect(pegoutSpy).toHaveBeenCalledTimes(1)
    expect(pegoutSpy).toHaveBeenNthCalledWith(1, {
      swap: expect.any(pTokensSwap),
      ptokenFrom: from,
      ptokenTo: to,
      dispatch,
    })
  })

  it('Should peg-in ETH on EOS', async () => {
    const peginWithWalletSpy = vi.spyOn(peginWithWallet, 'default').mockResolvedValue()
    const addDestinationAssetSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'addDestinationAsset')
    const setAmountSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'setAmount')
    const from = assets.find((_el) => _el.id === 'ZMT')
    const to = assets.find((_el) => _el.id === 'ZMT_ON_BSC_MAINNET')
    const amount = '1000'
    const address = '0xF39d30Fa570db7940e5b3A3e42694665A1449E4B'
    const f = swap(from, to, amount, address)
    const dispatch = vi.fn()
    await f(dispatch)
    expect(setAmountSpy).toHaveBeenCalledTimes(1)
    expect(setAmountSpy).toHaveBeenNthCalledWith(1, '1000')
    expect(addDestinationAssetSpy).toHaveBeenCalledTimes(1)
    expect(addDestinationAssetSpy).toHaveBeenNthCalledWith(
      1,
      expect.anything(),
      '0xF39d30Fa570db7940e5b3A3e42694665A1449E4B'
    )
    expect(peginWithWalletSpy).toHaveBeenCalledTimes(1)
    expect(peginWithWalletSpy).toHaveBeenNthCalledWith(1, {
      swap: expect.any(pTokensSwap),
      ptokenFrom: from,
      ptokenTo: to,
      dispatch,
    })
  })

  it('Should peg-out pBTC on Algorand', async () => {
    const pegoutSpy = vi.spyOn(pegout, 'default').mockResolvedValue()
    const addDestinationAssetSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'addDestinationAsset')
    const setCustomTransactionsSpy = vi.spyOn(pTokensAlgorandAsset.prototype, 'setCustomTransactions')
    const setAmountSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'setAmount')
    const from = assets.find((_el) => _el.id === 'PBTC_ON_ALGORAND_MAINNET')
    const to = assets.find((_el) => _el.id === 'BTC')
    const amount = '1000'
    const address = '19qn1TWUUxyhFvxwVxNfhF3XaL2eK4ya65'
    const f = swap(from, to, amount, address)
    const dispatch = vi.fn()
    await f(dispatch)
    expect(setAmountSpy).toHaveBeenCalledTimes(1)
    expect(setAmountSpy).toHaveBeenNthCalledWith(1, '1000')
    expect(addDestinationAssetSpy).toHaveBeenCalledTimes(1)
    expect(addDestinationAssetSpy).toHaveBeenNthCalledWith(1, expect.anything(), '19qn1TWUUxyhFvxwVxNfhF3XaL2eK4ya65')
    expect(pegoutSpy).toHaveBeenCalledTimes(1)
    expect(pegoutSpy).toHaveBeenNthCalledWith(1, {
      swap: expect.any(pTokensSwap),
      ptokenFrom: from,
      ptokenTo: to,
      dispatch,
    })
    expect(setCustomTransactionsSpy).toHaveBeenCalledTimes(0)
  })

  it('Should peg-out USDC on Algorand', async () => {
    vi.spyOn(algosdk.Algodv2.prototype, 'getTransactionParams').mockReturnValue({
      do: () =>
        Promise.resolve({
          flatFee: false,
          fee: 0,
          firstRound: 29759365,
          lastRound: 29760365,
          genesisID: 'mainnet-v1.0',
          genesisHash: 'wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=',
        }),
    })
    vi.spyOn(wallets, 'getWallets').mockReturnValue({
      algorand: { provider: undefined, account: 'VCMJKWOY5P5P7SKMZFFOCEROPJCZOTIJMNIYNUCKH7LRO45JMJP6UYBIJA' },
      eth: {},
    })
    const pegoutSpy = vi.spyOn(pegout, 'default').mockResolvedValue()
    const addDestinationAssetSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'addDestinationAsset')
    const setCustomTransactionsSpy = vi.spyOn(pTokensAlgorandAsset.prototype, 'setCustomTransactions')
    const setAmountSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'setAmount')
    const from = assets.find((_el) => _el.id === 'USDC_ON_ALGORAND_MAINNET')
    const to = assets.find((_el) => _el.id === 'USDC')
    const amount = '1000'
    const address = '0xF39d30Fa570db7940e5b3A3e42694665A1449E4B'
    const f = swap(from, to, amount, address)
    const dispatch = vi.fn()
    await f(dispatch)
    expect(setAmountSpy).toHaveBeenCalledTimes(1)
    expect(setAmountSpy).toHaveBeenNthCalledWith(1, '1000')
    expect(addDestinationAssetSpy).toHaveBeenCalledTimes(1)
    expect(addDestinationAssetSpy).toHaveBeenNthCalledWith(
      1,
      expect.anything(),
      '0xF39d30Fa570db7940e5b3A3e42694665A1449E4B'
    )
    expect(pegoutSpy).toHaveBeenCalledTimes(1)
    expect(pegoutSpy).toHaveBeenNthCalledWith(1, {
      swap: expect.any(pTokensSwap),
      ptokenFrom: from,
      ptokenTo: to,
      dispatch,
    })
    expect(setCustomTransactionsSpy).toHaveBeenCalledTimes(1)
    expect(setCustomTransactionsSpy).toMatchSnapshot()
  })

  it('Should peg-in USDC on Algorand', async () => {
    vi.spyOn(wallets, 'getWallets').mockReturnValue({
      eth: { account: '0xF39d30Fa570db7940e5b3A3e42694665A1449E4B' },
      algorand: { provider: undefined, account: 'VCMJKWOY5P5P7SKMZFFOCEROPJCZOTIJMNIYNUCKH7LRO45JMJP6UYBIJA' },
    })
    const peginWithWalletSpy = vi.spyOn(peginWithWallet, 'default').mockResolvedValue()
    const addDestinationAssetSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'addDestinationAsset')
    const setAmountSpy = vi.spyOn(pTokensSwapBuilder.prototype, 'setAmount')
    const from = assets.find((_el) => _el.id === 'USDC')
    const to = assets.find((_el) => _el.id === 'USDC_ON_ALGORAND_MAINNET')
    const amount = '1000'
    const address = 'VCMJKWOY5P5P7SKMZFFOCEROPJCZOTIJMNIYNUCKH7LRO45JMJP6UYBIJA'
    const f = swap(from, to, amount, address)
    const dispatch = vi.fn()
    await f(dispatch)
    expect(setAmountSpy).toHaveBeenCalledTimes(1)
    expect(setAmountSpy).toHaveBeenNthCalledWith(1, '1000')
    expect(addDestinationAssetSpy).toHaveBeenCalledTimes(1)
    expect(addDestinationAssetSpy).toHaveBeenNthCalledWith(
      1,
      expect.anything(),
      '770102986',
      '0x94c44b95c420a8989559d8ebfaffc94cc94ae1122e7a45974d09635186d04a3fd71773a9625fc408000000002c98cbf1c408000000003b9aca00c4080000000001e1ab70c408000000000000000091c420a8989559d8ebfaffc94cc94ae1122e7a45974d09635186d04a3fd71773a9625f9092ce2c98cbf1ce01e1ab70'
    )
    expect(peginWithWalletSpy).toHaveBeenCalledTimes(1)
    expect(peginWithWalletSpy).toHaveBeenNthCalledWith(1, {
      swap: expect.any(pTokensSwap),
      ptokenFrom: from,
      ptokenTo: to,
      dispatch,
    })
  })
})
