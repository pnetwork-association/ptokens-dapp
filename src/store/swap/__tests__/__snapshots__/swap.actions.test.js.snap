// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`swap > Should peg-out USDC on Algorand 1`] = `
[MockFunction setCustomTransactions] {
  "calls": [
    [
      [
        {
          "aamt": 1000000000,
          "arcv": {
            "data": [
              89,
              14,
              227,
              159,
              254,
              158,
              136,
              51,
              112,
              24,
              249,
              209,
              154,
              142,
              255,
              175,
              171,
              244,
              202,
              172,
              20,
              255,
              83,
              19,
              103,
              154,
              181,
              176,
              144,
              104,
              66,
              201,
            ],
            "type": "Buffer",
          },
          "fee": 1000,
          "fv": 29759365,
          "gen": "mainnet-v1.0",
          "gh": {
            "data": [
              192,
              97,
              196,
              216,
              252,
              29,
              189,
              222,
              210,
              215,
              96,
              75,
              228,
              86,
              142,
              63,
              109,
              4,
              25,
              135,
              172,
              55,
              189,
              228,
              182,
              32,
              181,
              171,
              57,
              36,
              138,
              223,
            ],
            "type": "Buffer",
          },
          "lv": 29760365,
          "note": {
            "data": [
              148,
              0,
              148,
              0,
              95,
              204,
              231,
              204,
              249,
              217,
              42,
              48,
              120,
              70,
              51,
              57,
              100,
              51,
              48,
              70,
              97,
              53,
              55,
              48,
              100,
              98,
              55,
              57,
              52,
              48,
              101,
              53,
              98,
              51,
              65,
              51,
              101,
              52,
              50,
              54,
              57,
              52,
              54,
              54,
              53,
              65,
              49,
              52,
              52,
              57,
              69,
              52,
              66,
              144,
            ],
            "type": "Buffer",
          },
          "snd": {
            "data": [
              168,
              152,
              149,
              89,
              216,
              235,
              250,
              255,
              201,
              76,
              201,
              74,
              225,
              18,
              46,
              122,
              69,
              151,
              77,
              9,
              99,
              81,
              134,
              208,
              74,
              63,
              215,
              23,
              115,
              169,
              98,
              95,
            ],
            "type": "Buffer",
          },
          "type": "axfer",
          "xaid": 31566704,
        },
        {
          "apaa": [
            {
              "data": [
                115,
                119,
                97,
                112,
              ],
              "type": "Buffer",
            },
            {
              "data": [
                0,
                0,
                0,
                0,
                44,
                152,
                203,
                241,
              ],
              "type": "Buffer",
            },
            {
              "data": [
                189,
                216,
                229,
                179,
                56,
                15,
                100,
                197,
                55,
                47,
                227,
                219,
                194,
                198,
                61,
                37,
                253,
                192,
                166,
                145,
                30,
                71,
                137,
                66,
                139,
                101,
                229,
                217,
                7,
                125,
                228,
                134,
              ],
              "type": "Buffer",
            },
          ],
          "apas": [
            748211185,
            31566704,
          ],
          "apat": [
            {
              "data": [
                189,
                216,
                229,
                179,
                56,
                15,
                100,
                197,
                55,
                47,
                227,
                219,
                194,
                198,
                61,
                37,
                253,
                192,
                166,
                145,
                30,
                71,
                137,
                66,
                139,
                101,
                229,
                217,
                7,
                125,
                228,
                134,
              ],
              "type": "Buffer",
            },
          ],
          "apid": 770102986,
          "fee": 1000,
          "fv": 29759365,
          "gen": "mainnet-v1.0",
          "gh": {
            "data": [
              192,
              97,
              196,
              216,
              252,
              29,
              189,
              222,
              210,
              215,
              96,
              75,
              228,
              86,
              142,
              63,
              109,
              4,
              25,
              135,
              172,
              55,
              189,
              228,
              182,
              32,
              181,
              171,
              57,
              36,
              138,
              223,
            ],
            "type": "Buffer",
          },
          "lv": 29760365,
          "snd": {
            "data": [
              168,
              152,
              149,
              89,
              216,
              235,
              250,
              255,
              201,
              76,
              201,
              74,
              225,
              18,
              46,
              122,
              69,
              151,
              77,
              9,
              99,
              81,
              134,
              208,
              74,
              63,
              215,
              23,
              115,
              169,
              98,
              95,
            ],
            "type": "Buffer",
          },
          "type": "appl",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": pTokensAlgorandAsset {
        "_assetInfo": {
          "chainId": "0x03c38e67",
          "decimals": 6,
          "fees": {
            "basisPoints": {
              "hostToHost": 10,
              "hostToNative": 25,
            },
            "minNodeOperatorFee": 10000000000000000000,
            "minNodeOperatorFeeUsd": 10,
            "networkFee": 0,
            "networkFeeUsd": 0,
          },
          "identity": "XXMOLMZYB5SMKNZP4PN4FRR5EX64BJURDZDYSQULMXS5SB354SDLBZX6HI",
          "isNative": false,
          "tokenAddress": "748211185",
          "tokenDecimals": 6,
          "tokenReference": "0xc8bb450f6c03e38ff83e01bc46fc2d24cb270f16",
          "vaultAddress": null,
        },
        "_customTransactions": [
          Transaction {
            "amount": 1000000000,
            "assetIndex": 31566704,
            "fee": 1000,
            "firstRound": 29759365,
            "from": {
              "checksum": Uint8Array [
                234,
                96,
                40,
                72,
              ],
              "publicKey": Uint8Array [
                168,
                152,
                149,
                89,
                216,
                235,
                250,
                255,
                201,
                76,
                201,
                74,
                225,
                18,
                46,
                122,
                69,
                151,
                77,
                9,
                99,
                81,
                134,
                208,
                74,
                63,
                215,
                23,
                115,
                169,
                98,
                95,
              ],
            },
            "genesisHash": {
              "data": [
                192,
                97,
                196,
                216,
                252,
                29,
                189,
                222,
                210,
                215,
                96,
                75,
                228,
                86,
                142,
                63,
                109,
                4,
                25,
                135,
                172,
                55,
                189,
                228,
                182,
                32,
                181,
                171,
                57,
                36,
                138,
                223,
              ],
              "type": "Buffer",
            },
            "genesisID": "mainnet-v1.0",
            "lastRound": 29760365,
            "lease": Uint8Array [],
            "name": "Transaction",
            "note": Uint8Array [
              148,
              0,
              148,
              0,
              95,
              204,
              231,
              204,
              249,
              217,
              42,
              48,
              120,
              70,
              51,
              57,
              100,
              51,
              48,
              70,
              97,
              53,
              55,
              48,
              100,
              98,
              55,
              57,
              52,
              48,
              101,
              53,
              98,
              51,
              65,
              51,
              101,
              52,
              50,
              54,
              57,
              52,
              54,
              54,
              53,
              65,
              49,
              52,
              52,
              57,
              69,
              52,
              66,
              144,
            ],
            "tag": {
              "data": [
                84,
                88,
              ],
              "type": "Buffer",
            },
            "to": {
              "checksum": Uint8Array [
                104,
                63,
                210,
                5,
              ],
              "publicKey": Uint8Array [
                89,
                14,
                227,
                159,
                254,
                158,
                136,
                51,
                112,
                24,
                249,
                209,
                154,
                142,
                255,
                175,
                171,
                244,
                202,
                172,
                20,
                255,
                83,
                19,
                103,
                154,
                181,
                176,
                144,
                104,
                66,
                201,
              ],
            },
            "type": "axfer",
          },
          Transaction {
            "appAccounts": [
              {
                "checksum": Uint8Array [
                  176,
                  230,
                  254,
                  58,
                ],
                "publicKey": Uint8Array [
                  189,
                  216,
                  229,
                  179,
                  56,
                  15,
                  100,
                  197,
                  55,
                  47,
                  227,
                  219,
                  194,
                  198,
                  61,
                  37,
                  253,
                  192,
                  166,
                  145,
                  30,
                  71,
                  137,
                  66,
                  139,
                  101,
                  229,
                  217,
                  7,
                  125,
                  228,
                  134,
                ],
              },
            ],
            "appArgs": [
              Uint8Array [
                115,
                119,
                97,
                112,
              ],
              Uint8Array [
                0,
                0,
                0,
                0,
                44,
                152,
                203,
                241,
              ],
              Uint8Array [
                189,
                216,
                229,
                179,
                56,
                15,
                100,
                197,
                55,
                47,
                227,
                219,
                194,
                198,
                61,
                37,
                253,
                192,
                166,
                145,
                30,
                71,
                137,
                66,
                139,
                101,
                229,
                217,
                7,
                125,
                228,
                134,
              ],
            ],
            "appForeignAssets": [
              748211185,
              31566704,
            ],
            "appIndex": 770102986,
            "fee": 1000,
            "firstRound": 29759365,
            "from": {
              "checksum": Uint8Array [
                234,
                96,
                40,
                72,
              ],
              "publicKey": Uint8Array [
                168,
                152,
                149,
                89,
                216,
                235,
                250,
                255,
                201,
                76,
                201,
                74,
                225,
                18,
                46,
                122,
                69,
                151,
                77,
                9,
                99,
                81,
                134,
                208,
                74,
                63,
                215,
                23,
                115,
                169,
                98,
                95,
              ],
            },
            "genesisHash": {
              "data": [
                192,
                97,
                196,
                216,
                252,
                29,
                189,
                222,
                210,
                215,
                96,
                75,
                228,
                86,
                142,
                63,
                109,
                4,
                25,
                135,
                172,
                55,
                189,
                228,
                182,
                32,
                181,
                171,
                57,
                36,
                138,
                223,
              ],
              "type": "Buffer",
            },
            "genesisID": "mainnet-v1.0",
            "lastRound": 29760365,
            "lease": Uint8Array [],
            "name": "Transaction",
            "note": Uint8Array [],
            "tag": {
              "data": [
                84,
                88,
              ],
              "type": "Buffer",
            },
            "type": "appl",
          },
        ],
        "_node": pTokensNode {
          "_provider": pTokensNodeProvider {},
        },
        "_provider": pTokensAlgorandProvider {
          "_account": "VCMJKWOY5P5P7SKMZFFOCEROPJCZOTIJMNIYNUCKH7LRO45JMJP6UYBIJA",
          "_client": AlgodClient {
            "c": HTTPClient {
              "bc": URLTokenBaseHTTPClient {
                "baseURL": "https://mainnet-api.algonode.cloud/",
                "defaultHeaders": {},
                "tokenHeader": {},
              },
            },
            "intDecoding": "default",
          },
          "_signer": undefined,
        },
        "_symbol": "USDC",
        "_type": 3,
        "_weight": 1,
      },
    },
  ],
}
`;
